src/CVRF/cvrf_eval.c

```diff
@@ -0,0 +1,423 @@
+	if (source == NULL)
+		return NULL;
+
+	struct cvrf_session *ret = oscap_alloc(sizeof(struct cvrf_session));
+	ret->source = source;
+	ret->model = cvrf_model_import(source);
+	ret->os_name = NULL;
+	ret->export_file = NULL;
+	ret->results_file = NULL;
+	ret->product_ids = oscap_stringlist_new();
+	ret->def_model = oval_definition_model_new();
+	return ret;
+}
+
+void cvrf_session_free(struct cvrf_session *session) {
+	if (session == NULL)
+		return;
+
+	cvrf_model_free(session->model);
+	oscap_free(session->os_name);
+	oscap_source_free(session->source);
+	oscap_free(session->export_file);
+	oscap_free(session->results_file);
+	oscap_stringlist_free(session->product_ids);
+	oval_definition_model_free(session->def_model);
+	oscap_free(session);
+}
+
+
+struct cvrf_rpm_attributes {
+	char *full_package_name;
+	char *rpm_name;
+	char *evr_format;
+};
+OSCAP_ACCESSOR_STRING(cvrf_rpm_attributes, full_package_name)
+OSCAP_ACCESSOR_STRING(cvrf_rpm_attributes, rpm_name)
+OSCAP_ACCESSOR_STRING(cvrf_rpm_attributes, evr_format)
+
+
+struct cvrf_rpm_attributes *cvrf_rpm_attributes_new() {
+	struct cvrf_rpm_attributes *ret = oscap_alloc(sizeof(struct cvrf_rpm_attributes));
+	if (ret == NULL)
+		return NULL;
+
+	ret->full_package_name = NULL;
+	ret->rpm_name = NULL;
+	ret->evr_format = NULL;
+	return ret;
+}
+
+void cvrf_rpm_attributes_free(struct cvrf_rpm_attributes *attributes) {
+	if (attributes == NULL)
+		return;
+
+	oscap_free(attributes->full_package_name);
+	oscap_free(attributes->rpm_name);
+	oscap_free(attributes->evr_format);
+	oscap_free(attributes);
+}
+
+/*
+ * End of structure definitions
+ *****************************************************************************/
+
+
+#define TAG_CVRF_DOC BAD_CAST "cvrfdoc"
+#define TAG_DOC_TITLE BAD_CAST "DocumentTitle"
+#define TAG_DOC_TYPE BAD_CAST "DocumentType"
+#define ATTR_PRODUCT_ID "ProductID"
+//namespaces
+#define CVRF_NS BAD_CAST "http://www.icasi.org/CVRF/schema/cvrf/1.1"
+#define VULN_NS BAD_CAST "http://www.icasi.org/CVRF/schema/vuln/1.1"
+
+static int find_all_cvrf_product_ids_from_cpe(struct cvrf_session *session) {
+	if (cvrf_model_filter_by_cpe(session->model, session->os_name) == -1)
+		return -1;
+
+	struct cvrf_product_tree *tree = cvrf_model_get_product_tree(session->model);
+	struct cvrf_relationship_iterator *relationships = cvrf_product_tree_get_relationships(tree);
+	while (cvrf_relationship_iterator_has_more(relationships)) {
+		struct cvrf_relationship *relation = cvrf_relationship_iterator_next(relationships);
+		struct cvrf_product_name *name = cvrf_relationship_get_product_name(relation);
+		oscap_stringlist_add_string(session->product_ids, cvrf_product_name_get_product_id(name));
+	}
+	cvrf_relationship_iterator_free(relationships);
+	return 0;
+}
+
+int cvrf_export_results(struct oscap_source *import_source, const char *export_file, const char *os_name) {
+	__attribute__nonnull__(import_source);
+	__attribute__nonnull__(export_file);
```

Comment:
This is just a temporary way of providing the CPE name. But it cannot be null, otherwise we cannot do any kind of vulnerability checking since we won't know which products/ProductIDs to look at